client/api_handler.py

```diff
@@ -24,55 +25,67 @@
     desc_length = query_params.desc_length
     response_type = query_params.response_type
-
-    logger = current_app.logger
-    logger.setLevel(logging.DEBUG)
-    logger.debug("FUNC: get_metadata, url: %s, response_type: %s" % (url, response_type))
-    sanitized_url = url_utils.sanitize_url(url)
-    logger.debug("Sanitized url: %s" % sanitized_url)
-
-    metadata = None
-    content_type = None
-    response_code = None
-
-    head = url_utils.get_requests_header(sanitized_url)
-    if head is None:
-        metadata = create_metadata_object(url, StatusCode.BAD_REQUEST, None, None)
-        response_code = StatusCode.BAD_REQUEST
-        sanitized_url = url
+    json_output = collections.OrderedDict()
+    json_output['length'] = len(urls)
+    json_output['responses'] = []
+
+    for url in urls:
+        url = url.strip()
+        logger = current_app.logger
+        logger.setLevel(logging.DEBUG)
+        logger.debug("FUNC: get_metadata, url: %s, response_type: %s" % (url, response_type))
+        sanitized_url = url_utils.sanitize_url(url)
+        logger.debug("Sanitized url: %s" % sanitized_url)
+
+        metadata = None
+        content_type = None
+        response_code = None
+
+        head = url_utils.get_requests_header(sanitized_url)
+        if head is None:
+            metadata = create_metadata_object(url, StatusCode.BAD_REQUEST, None, None)
+            response_code = StatusCode.BAD_REQUEST
+            sanitized_url = url
+        else:
+            sanitized_url = url_utils.get_redirect_url(head)
+            content_type = url_utils.get_content_type(head)
+            response_code = head.status_code
+            metadata = create_metadata_object(url, response_code, sanitized_url, content_type)
+
+        site_response = metadata.fetch_site_data(sanitized_url, response_code)
+
+        # Check for caching, otherwise proceed with scraping
+        if config.CACHE_DATA:
+            data = cache_utils.get_cached_data(sanitized_url)   # If data from db is None, continue and parse the website
+            if data is not None:
+                logger.debug("Cache hit for key %s", sanitized_url)
+                data_map = json.loads(data.metadata)
+                metadata.data_map[FieldKeyword.DATA] = data_map
+                return get_json_metadata(metadata.data_map)
+
+        metadata.parse_content(site_response)
+
+        # Trim the description if necessary
+        trim_description(metadata, desc_length)
+
+        json_data = get_json_metadata(metadata.data_map)
+        logger.info(json_data)
+        if config.CACHE_DATA and site_response.status_code == StatusCode.OK:
+            cache_map = metadata.get_cache_prop_map()
+            cache_data = get_json_metadata(cache_map)
+            logger.debug("Caching json data to redis db")
+            cache_utils.cache_json_data(sanitized_url, cache_data)
+        json_output['responses'].append(json.loads(json_data))
+
+    if json_output['length'] > 1:
```

Comment:
I think this is a little messy, can we extract out the getting multiple urls & getting the metadata for one url to a different function?
eg:
```def get_url_metadata(query_params):
...

def get_urls_metadata(query_params):
< logic for getting many responses and making into a json call get_url_metadata>

```
In this case, we can easily make `get_url_metadata` a concurrent function.
Also, in order to make it possible to be concurrent you should make query_params an array of QueryParam objects rather than have urls in query_param.